### ISSUE TITLE
Need more complete examples

### ISSUE THREAD

**Issue by mgdodge** [paleo/direct-vuex#2]
I am trying to implement direct-vuex in a 'boilerplate' I frequently use when setting up new projects. It is written in typescript with vuex-typex, but direct-vuex is a lot cleaner and closer to the vuex docs, so I would like to switch to it.

However, the docs here are unclear in a couple of areas, specifically with regards to actions. These are the last hurdles I have to clear in the transition.

The first issue:

> Actions can't be declared with the object alternative syntax.

I am not sure what this means - I don't fully know what "alternative" means here. Could you give an example of bad/good syntax?

Also, I cannot figure out how to properly use context in vuex (actions, etc.)  A sample for writing/using a vuex module would be helpful. For example, from the vuex docs:

```javascript
const moduleA = {
  // ...
  actions: {
    incrementIfOddOnRootSum ({ state, commit, rootState }) {
      if ((state.count + rootState.count) % 2 === 1) {
        commit('increment')
      }
    }
  }
}
```

Note that the first argument of the action is the _context_ which in this example contains the module state, module commit, and root state. Likewise, in the root store, the context would have similar properties.

It seems this context argument is still required, but I can't seem to get it to work. Can you please provide an example of how to get type hinting/autocompletion on the context-specific state/commit/etc. via direct-vuex?

If direct-vuex can't get that type information, the argument is still required in actions, but is no longer useful, so it should be communicated that you can use store.state and store.commit instead.

**Comment by paleo** [comment]
> I don't fully know what "alternative" means here. Could you give an example of bad/good syntax?

I found an alternative syntax [in the types of Vuex](https://github.com/vuejs/vuex/blob/dev/types/index.d.ts#L100):

```
export type ActionHandler<S, R> = (this: Store<R>, injectee: ActionContext<S, R>, payload?: any) => any;
export interface ActionObject<S, R> {
  root?: boolean;
  handler: ActionHandler<S, R>;
}
export type Action<S, R> = ActionHandler<S, R> | ActionObject<S, R>;
```

So, an `Action` can be an `ActionHandler` (which is a function) or an `ActionObject` that contains the handler and a `root` property. But I did not found any official documentation about the `ActionObject` syntax. Maybe I should just drop this paragraph from the README file.

> I cannot figure out how to properly use context in vuex (actions, etc.) A sample for writing/using a vuex module would be helpful.

There is almost no difference. Direct-vuex doesn't provide any help in the store implementation. So, the example from the Vuex documentation is valid. But you'll have to append `as const` at the end of the object declaration: 

```
const moduleA = {
  // …
} as const
```

Direct-vuex provides typed wrappers _for the code that uses the store_. Not for the implementation of the store itself. I wrote [an article in Medium](https://itnext.io/use-a-vuex-store-with-typing-in-typescript-without-decorators-or-boilerplate-57732d175ff3) with a more detailed example.

**Comment by mgdodge** [comment]
Thank you for the explanation of the alternate syntax. I thought I was missing something.

As for the question about context, I know that this is just a wrapper, and that I should write things the way I normally would...but the problem is that typescript complains. If I have:

```javascript
incrementIfOddOnRootSum ({ state, commit, rootState }) {
...
```

My IDE complains that `state`, `commit`, and `rootState` implicitly have type 'any' and therefore have no autocompletion/type hinting in them.

When using the store in vue components, direct-vuex does an excellent job providing all the typings that would be needed while using a syntax that is as close to vanilla vuex as I have ever seen. This is wonderful!

It seems to fall short, though, with the context argument for actions. This is important, as mutations are almost always committed from within actions (in my experience), and proper typing of mutation payloads is most often encountered here. With the example above, I can't get autocompletion on `state.<key>` when trying to check existing values, nor on `commit.<method>` when I need to mutate the data, which by extension means I won't be able to get type checking on the payload I send to that commit.

I have discovered a workaround for the time being that seems to work well, and provides typings as needed inside actions. In essence, the context is still passed, but remains unused, and state/commit/etc. are accessed via the fully typed root store. For example:

```typescript
// --- store.ts
import Vue from "vue"
import Vuex from "vuex"
import { createDirectStore } from "direct-vuex"
import module1 from "./module1"

Vue.use(Vuex)

interface RootState {
    foo: string
    bar: null | any
}

const store = createDirectStore({
  state: {
    foo: '',
    bar: null
  },
  mutations: {
    SET_BAR(state: RootState, data: BarData) { ... }
  },
  actions: {
    getBar(_context: any) {
      const state = store.state
      const commit = store.commit
      if (!state.bar) {
        fetch('/someurl').then((data) => {
          commit.SET_BAR(data)
        })
      } else {
        // do something else
      }
    }
  },
  modules: {
    module1
  }
} as const)

export default store
export type AppStore = typeof store

declare module "vuex" {
  interface Store<S> {
    direct: AppStore
  }
}






// --- module1
import store from './store.ts'

export interface Module1State {
  name: null | string
}

export default {
  namespaced: true,
  state: {
    name: null
  } as Module1State,
  getters: {
    message: (state: Module1State) => `Hello, ${state.name}!`
  },
  mutations: {
    SET_NAME(state: Module1State, newName: string) {
      state.name = newName
    },
  },
  actions: {
    async loadName(_context: any, payload: { id: string }) {
      const state = store.state.module1
      const commit = store.commit.module1
      let name = null
      // Only load name from somewhere if not already set
      if (!state.name) {
        name = `Name-${payload.id}` // load it from somewhere
        commit.SET_NAME(name)
      } else {
        name = state.name;
      }
      return { name }
    },
  }
} as const
```

Note that in both the root store actiosn as well as the module actions, types are available via the root store. I have tested this in the boilerplate project that I have, and the only warnings I got were because of circular references, which can be safely ignored in this usage.

I am open to other better solutions, but this seemed the best way to get typings inside of actions/etc. with minimal jumping through hoops. In any case, I think it would be good to either document this option, or the fact that the context will be 'any' typed. And a more complete example would illustrate these. Thoughts?

**Comment by paleo** [comment]
You could use the type from Vuex, which is far from optimal, but this should make the error with implicit type `any` disappear.

```ts
import { ActionContext } from "vuex"
incrementIfOddOnRootSum ({ state, commit, rootState }: ActionContext<any, any>) {
}
```

Or in the module:

```ts
  actions: {
    async loadName(_context: ActionContext<Module1State, any>, payload: { id: string }) {
```

Personally I don't like the circular reference. Maybe it is possible to find a good solution but I didn't find it for the moment so I continue to implement the store without automatic typing. I use the trick of `Module1State` for the state, and that's all.

**Comment by paleo** [comment]
I tried something for the context of action handlers. A minimal documentation [is here](https://github.com/paleo/direct-vuex#use-typed-wrappers-from-action-implementations).
@mgdodge

**Comment by mgdodge** [comment]
Anxious to try this out! I will try updating my boilerplate with these changes, and will update this issue once I've given it a go. If this works, I think it is the most elegant typescript-vuex solution so far - thanks!

**Comment by mgdodge** [comment]
Closing this issue - the [new feature and accompanying documentation](https://github.com/paleo/direct-vuex#use-typed-wrappers-from-action-implementations), mentioned [here](https://github.com/paleo/direct-vuex/issues/2#issuecomment-545157732) and implemented in v0.3.3, adequately address the issue.

**Comment by paleo** [comment]
@mgdodge There is a performance issue with Action Handlers in version 0.3.3. I've just published the version 0.5.0. Be careful, creating a store is a little bit different:

    const { store, directActionContext, directRootActionContext } = createDirectStore(…)
    export default store
    export { directActionContext, directRootActionContext }

Other breaking change: the order of parameters is inversed in `directActionContext`. Now it is:

    directActionContext(context, myModule)

**Comment by ilanbenyos** [comment]
@mgdodge is right. I would also appreciate a full working store with modules, rootgetters, rootstore ect.
